src/main/java/io/confluent/kafkarest/ConsumerReadTask.java

```diff
@@ -108,37 +110,50 @@ public boolean doPartialRead() {
       long startedIteration = parent.getConfig().getTime().milliseconds();
       final int requestTimeoutMs = parent.getConfig().getInt(
           KafkaRestConfig.CONSUMER_REQUEST_TIMEOUT_MS_CONFIG);
+      final long endTime = startedIteration + requestTimeoutMs;
+      final int itBackoff = parent.getConfig().getInt(
+        KafkaRestConfig.CONSUMER_ITERATOR_BACKOFF_MS_CONFIG);
       try {
-        // Read off as many messages as we can without triggering a timeout exception. The
-        // consumer timeout should be set very small, so the expectation is that even in the
-        // worst case, num_messages * consumer_timeout << request_timeout, so it's safe to only
-        // check the elapsed time once this loop finishes.
-        while (iter.hasNext()) {
-          MessageAndMetadata<KafkaK, KafkaV> msg = iter.peek();
-          ConsumerRecordAndSize<ClientK, ClientV> recordAndSize = parent.createConsumerRecord(msg);
-          long roughMsgSize = recordAndSize.getSize();
-          if (bytesConsumed + roughMsgSize > maxResponseBytes) {
-            break;
+        while (parent.getConfig().getTime().milliseconds() < endTime) {
+
+          if (iter == null || !iter.hasNext()) {
+            // The consumer timeout should be set very small, so the expectation is that even in the
+            // worst case, num_messages * consumer_timeout << request_timeout, so it's safe to only
+            // check the elapsed time once this loop finishes.
+            ConsumerRecords<KafkaK, KafkaV> records = consumer.poll(itBackoff);
```

Comment:
I think we might need to think through this a bit. The old consumer did most of its work (including rebalancing and fetching) in the background, but the new consumer is designed to work as an event loop with the user calling poll() repeatedly and all of the work done in the foreground inside poll(). When you stop calling poll(), the consumer stops sending heartbeats and the broker reassigns its partitions to another member. So I think we may need to do a little more work to ensure that consumers stay in the group as long as users are fetching data. 

Thinking outloud... One option I can think of would be to change `ConsumerWorker` to work only with a single instance of `KafkaConsumer`. Since each worker gets its own thread, it can execute the poll loop for that consumer instance and feed records into a queue, which can then be accessed asynchronously. Then we could have `ConsumerManager` maintain a map from the `ConsumerInstanceId` to the `ConsumerWorker`. This might even simplify expiration since we can have the worker itself shutdown when we see no reads from the consumer after some time.